src/pykeen/triples/splitting.py

```diff
@@ -275,6 +255,194 @@ def _prepare_cleanup(
+
+        :return:
+            a sequence of ID-based triples for each split part. the absolute may be different to ensure the constraint.
+        """
+        raise NotImplementedError
+
+    def split(
+        self,
+        *,
+        mapped_triples: MappedTriples,
+        ratios: Union[float, Sequence[float]] = 0.8,
+        random_state: TorchRandomHint = None,
+    ) -> Sequence[MappedTriples]:
+        """Split triples into clean groups.
+
+        :param mapped_triples: shape: (n, 3)
+            the ID-based triples
+        :param random_state:
+            the random state used to shuffle and split the triples
+        :param ratios:
+            There are three options for this argument.
+            First, a float can be given between 0 and 1.0, non-inclusive. The first set of triples will get this
+            ratio and the second will get the rest.
+            Second, a list of ratios can be given for which set in which order should get what ratios as
+            in ``[0.8, 0.1]``.
+            The final ratio can be omitted because that can be calculated.
+            Third, all ratios can be explicitly set in order such as in ``[0.8, 0.1, 0.1]`` where the sum of
+            all ratios is 1.0.
+
+        :return:
+            A partition of triples, which are split (approximately) according to the ratios.
+        """
+        random_state = ensure_torch_random_state(random_state)
+        ratios = normalize_ratios(ratios=ratios)
+        sizes = get_absolute_split_sizes(n_total=mapped_triples.shape[0], ratios=ratios)
+        triples_groups = self.split_absolute_size(
+            mapped_triples=mapped_triples,
+            sizes=sizes,
+            random_state=random_state,
+        )
+        for i, (triples, exp_size, exp_ratio) in enumerate(zip(triples_groups, sizes, ratios)):
+            actual_size = triples.shape[0]
+            actual_ratio = actual_size / exp_size * exp_ratio
+            if actual_size != exp_size:
+                logger.warning(
+                    f"Requested ratio[{i}]={exp_ratio:.3f} (equal to size {exp_size}), but got {actual_ratio:.3f} "
+                    f"(equal to size {actual_size}) to ensure that all entities/relations occur in train.",
+                )
+        return triples_groups
+
+
+class CleanupSplitter(Splitter):
+    """
+    The cleanup splitter first randomly splits the triples and then cleans up.
+
+    In the cleanup process, triples are moved into the train part until all entities occur at least once in train.
+
+    The splitter supports two variants of cleanup, cf. ``cleaner_resolver``.
+    """
+
+    def __init__(self, cleaner: HintOrType[Cleaner] = None) -> None:
+        """
+        Initialize the splitter.
+
+        :param cleaner:
+            the cleanup method to use. Defaults to the fast deterministic cleaner,
+            which may lead to larger deviances between desired and actual triple count.
+        """
+        super().__init__()
```

Comment:
not necessary since no init defined in base class

Reply:
5cc5f90